Environment support

annettejanewilson opened the pull request:
Add support for environment-aware commands. Specify --environment-id with most commands on content-type, entries, assets, locales.
Also adds missing locale, role and space-membership commands.

annettejanewilson committed:
Add basic environment support

annettejanewilson committed:
Add locale commands

annettejanewilson committed:
Add space membership commands

annettejanewilson committed:
Add role commands

annettejanewilson committed:
Update README with all supported commands

unlobito commented on contentful_cli/management.py:
```diff
@@ -393,7 +419,12 @@ def construct_base_url(self):
+
+        templating_arguments = dict(arguments)
+        if self.environment_aware and arguments.get('environment_id'):
+            templating_arguments['space_id'] += '/environments/' + arguments['environment_id']
```
This feels super hacky. Instead of modifying the endpoint templates (which had to be modified to flag them as environment capable anyway), this hotpatches another unrelated field to add the environment. It makes a possibly dangerous assumption about the ordering of the fields in the endpoint URL and could break with future endpoints.

We could make this clearer while maintaining the ability to not specify an environment in API calls (for accounts that haven't been migrated) by adding a `{environment}` templating argument that is `""` when an environment isn't used/set, but is `'/environments/' + arguments['environment_id']` when it is used.

annettejanewilson replied:
I was uncomfortable with this too. I considered what you suggest here, but backed off because it made the template strings extremely unwieldy without clear benefit.

I disagree that the space id is an "unrelated" field - the space and the environment are inextricably bound together. Also I don't see how the assumption about field order is "dangerous". But I would concede that the way it's used does not make it clear that it is a special case. How would you feel about replacing this with something that makes it clear it is getting special treatment, such as replacing this:

`/spaces/{space_id}/entries/{entry_id}`

With:

`{environment_specifier}/entries/{entry_id}`

Where we document `{environment_specifier}` as being one of:

```
spaces/{space_id}
spaces/{space_id}/environments/{environment_id}
```

...dependent on whether the endpoint is flagged as environment-aware and whether an environment is specified?

(Note also that it is *always* valid not to specify an environment for merely environment-aware commands. This implies use of the master environment. I think you're already clear on that but I want to mention it in case I'm not being clear myself.)

unlobito commented on contentful_cli/management.py:
```diff
@@ -393,7 +419,12 @@ def construct_base_url(self):
+        if self.environment_aware and arguments.get('environment_id'):
+            templating_arguments['space_id'] += '/environments/' + arguments['environment_id']
+
+        expanded_path = self.template.format(**templating_arguments)
```
Mixing `.format` and `+= calls` in the same block of code seems a bit confusing. We should stick to one method of string concatenation and use that consistently.

It seems `+=` is only used in [one other place](https://github.com/Skyscanner/skyscanner-contentful-management/blob/e17a8eaeba8a31995925d396d5e3bfadc3d08d6b/contentful_cli/management.py#L425), whereas `.format` is regularly used throughout the file (even for single concatenations).

annettejanewilson replied:
To be honest I don't see why `+=` and `format` preclude each other, but I'll change it if you prefer. I could see an argument that it's a bit messy to draw the space_id from the new dictionary and the environment_id from the old dictionary.

I'll revise this to try to make the intent a bit clearer.

unlobito commented on contentful_cli/management.py:
```diff
@@ -82,36 +82,57 @@ def echo_output(obj, echo_to_stdout, log_file):
+    Endpoint('put-role',                'put',    'api',    '/spaces/{space_id}/roles/{role_id}',                                  'ED'),
+    Endpoint('get-role',                'get',    'api',    '/spaces/{space_id}/roles/{role_id}',                                  'E'),
+    Endpoint('delete-role',             'delete', 'api',    '/spaces/{space_id}/roles/{role_id}',                                  'E'),
+    Endpoint('get-environment',         'get',    'api',    '/spaces/{space_id}/environments/{environment_id}',                    ''),
```
These are environment aware endpoints but don't use the same mechanism to specify the environment as other endpoints.

annettejanewilson replied:
I don't think I've been sufficiently clear here. The Contentful docs specifically describe "environment aware resources": http://ctf-doc-app-branch-environments.netlify.com/developers/docs/references/content-management-api/#environment-aware-resources

The environment resource itself is not "environment-aware" in this sense. The environment itself does not exist _inside_ an environment. I'll add a better definition of what "environment aware" means. Basically it's anything that exists inside an environment and that changing it in one environment will not affect another environment.

annettejanewilson committed:
Be explicit about special environment behaviour

Problem: The previous approach to handling environment-aware resources,
like entries, content_types and assets, obfuscated where and when the
environment ID would be inserted into the request URL, hiding it in the
space_id field. This could confuse anyone trying to maintain the code.

Solution: Add "environment_path" as a special path fragment avaiable in
the template when the environment-aware flag is enabled on an endpoint.
This will be "/spaces/{space_id}" when no environment is specified, and
"/spaces/{space_id}/environments/{environment_id}" if an environment is
specified.

annettejanewilson committed:
Simplify environment tests for readability

Extract out the common setup into fixtures.
